1410 xsched module

mn13 opened the pull request:
Closes https://github.com/warden-protocol/wardenprotocol/issues/1410

Example of callback:
```solidity
// SPDX-License-Identifier: UNLICENSED
pragma solidity ^0.8.25;

import "./precompiles/async/IAsync.sol";
import "./precompiles/sched/ISched.sol";
import "./precompiles/common/Types.sol";

error Unauthorized();
error NotReady();

contract CallbackExample {
    uint64 public lastTaskId;
    uint64 public lastCbId;

    bytes public output;

    function run() external {
        Types.Coin[] memory maxFee;
        lastTaskId = 
            IASYNC_CONTRACT.addTask("echo", hex"deadbeef", maxFee, CallbackParams(address(this), 100000000));

        TaskByIdResponse memory taskResponse = IASYNC_CONTRACT.taskById(lastTaskId);
        lastCbId = taskResponse.taskResponse.task.callbackId;
    }

    function t() external pure returns (bytes memory) {
        return hex"deadbeef";
    }

    function cb(uint64 id, bytes calldata o) external returns(bytes memory)  {
        if(msg.sender != ISCHED_CONTRACT.getAddress()) {
            revert Unauthorized();
        }

        if(id == lastCbId) {
            TaskByIdResponse memory task = IASYNC_CONTRACT.taskById(lastTaskId);

            if (task.taskResponse.result.id == 0) revert NotReady();

            output = task.taskResponse.result.output;
        }

        return o;
    }
}
```

mn13 committed:
scaffolding sched module

mn13 committed:
callback keeper

mn13 committed:
sched queries

mn13 committed:
add callback execution & app wiring

mn13 committed:
use x/sched callbacks in x/async

mn13 committed:
add callbackId field to Task in precompile

mn13 committed:
rename IAsyncCallback -> ICallback

mn13 committed:
sched precompile

mn13 committed:
fix test

mn13 commented on proto/warden/async/v1beta1/events.proto:
```diff
@@ -29,5 +29,5 @@
 
   // Address of callback that will be triggered after execution.
-  string callback_address = 4;
+  uint64 callback_id = 4;
```
I don't sure it's relevant while we are not in mainnet

mn13 commented on tests/justfile:
```diff
@@ -25,7 +25,7 @@ _rm_rf dir:
 
 # Define the new precompiles addresses
-warden-precompiles := '["0x0000000000000000000000000000000000000900", "0x0000000000000000000000000000000000000901", "0x0000000000000000000000000000000000000902", "0x0000000000000000000000000000000000000903", "0x0000000000000000000000000000000000000904"]'
+warden-precompiles := '["0x0000000000000000000000000000000000000900", "0x0000000000000000000000000000000000000901", "0x0000000000000000000000000000000000000902", "0x0000000000000000000000000000000000000903", "0x0000000000000000000000000000000000000904", "0x0000000000000000000000000000000000000905"]'
```
Registered at line 97

`precompiles[newSchedPrecompile.Address()] = newSchedPrecompile`

mn13 commented on warden/x/sched/keeper/keeper.go:
```diff
@@ -0,0 +1,268 @@
+			return nil
+		} else {
+			k.callbacks.setSucceed(ctx, id, res.Return())
+		}
```
I think 3rd party developer can't accurately estimate gas for estimation + execution.

mn13 committed:
format buf

mn13 committed:
fix rabbit

mn13 committed:
fix buf linter

mn13 commented on warden/x/sched/keeper/keeper.go:
```diff
@@ -0,0 +1,287 @@
+		return nil, err
+	}
+
+	if err := fromAcc.SetSequence(fromAcc.GetSequence() + 1); err != nil {
+		return nil, err
+	}
```
Ethereum bump nonce for failed transaction, I think we should do the same

mn13 committed:
fix golangci-lint

mn13 committed:
update changelog

mn13 committed:
fix rabbit

backsapc commented on precompiles/sched/Types.abi:
You suggested to delete this last time

backsapc commented on precompiles/sched/sched.go:
```diff
@@ -0,0 +1,138 @@
+//go:embed abi.json
+var f embed.FS
+
+// Precompile defines the precompiled contract for x/async.
```
```suggestion
// Precompile defines the precompiled contract for x/sched.
```

backsapc commented on precompiles/sched/sched.go:
```diff
@@ -0,0 +1,138 @@
+}
+
+// LoadABI loads the x/async ABI from the embedded abi.json file
+// for the x/async precompile.
```
```suggestion
// for the x/sched precompile.
```

backsapc commented on precompiles/sched/sched.go:
```diff
@@ -0,0 +1,138 @@
+	queryServer    types.QueryServer
+}
+
+// LoadABI loads the x/async ABI from the embedded abi.json file
```
```suggestion
// LoadABI loads the x/sched ABI from the embedded abi.json file
```

backsapc commented on precompiles/sched/types.go:
```diff
@@ -0,0 +1,116 @@
+	types "github.com/warden-protocol/wardenprotocol/warden/x/sched/types/v1beta1"
+)
+
+// FromResponse needed to map QueryTaskByIdResponse to TaskByIdResponse.
```
```suggestion
// FromResponse needed to map QueryCallbackByIdResponse to CallbackByIdResponse.
```

backsapc commented on precompiles/sched/types.go:
```diff
@@ -0,0 +1,116 @@
+// FromResponse needed to map QueryTaskByIdResponse to TaskByIdResponse.
+func (r *CallbackByIdResponse) FromResponse(res *types.QueryCallbackByIdResponse) (CallbackByIdResponse, error) {
+	if res != nil {
+		mappedTaskResponse, err := mapCallbackResponse(res.CallbackResponse)
```
```suggestion
		mappedCallbackResponse, err := mapCallbackResponse(res.CallbackResponse)
```

backsapc commented on precompiles/sched/types.go:
```diff
@@ -0,0 +1,116 @@
+			return CallbackByIdResponse{}, err
+		}
+
+		r.CallbackResponse = mappedTaskResponse
```
```suggestion
		r.CallbackResponse = mappedCallbackResponse
```

backsapc commented on warden/x/sched/keeper/keeper.go:
```diff
@@ -0,0 +1,287 @@
+	cbId uint64,
+) error {
+	feeAmt, fee := k.callbackFee(ctx, gas)
+	evmKeeper := k.getEvmKeeper(0)
```
Looks like a magic value even if you go to the func definition. Can we add constant? Comment? Give a better name to the parameter?

pitasi replied:
the underlying problem is that we need to fix our depinject configuration, right now this number is only being used to make the signature of the function getEvmKeeper slightly different

the actual value is not used

backsapc commented on warden/x/sched/keeper/keeper.go:
```diff
@@ -0,0 +1,287 @@
+		false,                 // isFake
+	)
+
+	evmKeeper := k.getEvmKeeper(0)
```
Same thing

mn13 committed:
review

pitasi commented on precompiles/common/Types.sol:
```diff
@@ -63,4 +63,9 @@ library Types {
         uint64 revisionHeight;
     }
+
+    struct CallbackParams {
```
i feel like this should be in the ISched contract, not in the common one

pitasi commented on proto/warden/sched/v1beta1/callback.proto:
```diff
@@ -0,0 +1,50 @@
+  CALLBACK_RESULT_SUCCEED = 1;
+  // Indicates that the callback execution failed.
+  CALLBACK_RESULT_FAILED = 2;
+}
```
yep, it'd be better to follow the same name, so "callback status" instead of "callback result"

pitasi commented on proto/warden/sched/v1beta1/query.proto:
```diff
@@ -0,0 +1,63 @@
+  // Parameters queries the parameters of the module.
+  rpc Params(QueryParamsRequest) returns (QueryParamsResponse) {
+    option (google.api.http).get = "/warden-protocol/wardenprotocol/sched/params";
+  }
```
:+1: i think we have been inconsistent in other parts of the protobufs, but it's best to start now 😁

pitasi commented on proto/warden/sched/v1beta1/query.proto:
```diff
@@ -0,0 +1,63 @@
+
+message QueryCallbackByIdResponse {
+  CallbackResponse callbackResponse = 1 [(gogoproto.nullable) = false];
+}
```
+1, let's follow snake case for proto fields

pitasi commented on warden/x/sched/keeper/keeper.go:
```diff
@@ -0,0 +1,289 @@
+		panic("invalid authority address: " + authority)
+	}
+
+	sb := collections.NewSchemaBuilder(storeService)
+
+	callbacks := NewCallbackKeeper(sb, cdc)
+
```
you should call `sb.Build()` and panic in case of errors (this happens for example if we have multiple collections with same prefix)

pitasi reviewed (changes requested):
looks fundamentally good, commenting on some minor things

mn13 committed:
review
        - format proto definitions
        - move CallbackParams into ISched.sol
        - add sb.Build() into keeper constructor

mn13 committed:
rebase fixes & fix panic when insufficient funds

mn13 committed:
fix solidity CI

pitasi commented on proto/warden/async/v1beta1/task.proto:
```diff
@@ -37,10 +37,12 @@ message Task {
   // Callback to be called when the Task is completed.
-  string callback = 5;
+  string callback = 5 [deprecated = true];
   // Deducted fee are the tokens collected this module when this Task was created.
```
since we don't handle this deprecation anywhere in the code, we should just remove the field altogether and replace it with a line:

```proto
reserved 5;
```

(which ensures we don't reuse the same field tag 5 again)

mn13 replied:
CI check not passed when using reserved (`buf breaking`) :(

pitasi commented on proto/warden/async/v1beta1/tx.proto:
```diff
@@ -63,14 +64,15 @@ message MsgAddTask {
   string plugin = 2;
   bytes input = 3;
-  string callback = 4;
+  string callback = 4 [deprecated = true];
```
replace with

```proto
reserved 4;
```

pitasi reviewed (changes requested):
final minor things, lgtm

mn13 committed:
use reserved for not used callback fields

pitasi reviewed (approved):
lgtm!
